### ISSUE TITLE
Factored Marian segfaults with guided alignments

### ISSUE THREAD

**Issue by tomsbergmanis** [marian-cef/marian-dev#4]
Factored Marian models work fine unless I add guided alignments. Let me know if you need more information. 
```
[2021-02-09 14:19:05] Error: Segmentation fault
[2021-02-09 14:19:05] Error: Aborted from setErrorHandlers()::<lambda(int, siginfo_t*, void*)> in [..]/software/marian_factors/marian-dev/src/common/logging.cpp:130

 

[CALL STACK]
[0x560ec97e3209]                                                       + 0x3a7209
[0x560ec97e35b9]                                                       + 0x3a75b9
[0x7f433d2668a0]                                                       + 0x128a0
[0x560ec98a2dbc]    marian::data::CorpusBase::  addAlignmentsToBatch  (std::shared_ptr<marian::data::CorpusBatch>,  std::vector<marian::data::SentenceTuple,std::allocator<marian::data::SentenceTuple>> const&) + 0x2fc
[0x560ec98bbed7]    marian::data::Corpus::  toBatch  (std::vector<marian::data::SentenceTuple,std::allocator<marian::data::SentenceTuple>> const&) + 0x1107
[0x560ec97ad673]    marian::data::BatchGenerator<marian::data::CorpusBase>::  fetchBatches  () + 0xa43
[0x560ec97aeb5d]    marian::ThreadPool::enqueue<marian::data::BatchGenerator<marian::data::CorpusBase>::fetchBatchesAsync()::{lambda()#1}>(std::result_of&&,(marian::data::BatchGenerator<marian::data::CorpusBase>::fetchBatchesAsync()::{lambda()#1}&&)...)::{lambda()#1}::  operator()  () const + 0x2d
[0x560ec97af8b0]    std::_Function_handler<std::unique_ptr<std::__future_base::_Result_base,std::__future_base::_Result_base::_Deleter> (),std::__future_base::_Task_setter<std::unique_ptr<std::__future_base::_Result<std::deque<std::shared_ptr<marian::data::CorpusBatch>,std::allocator<std::shared_ptr<marian::data::CorpusBatch>>>>,std::__future_base::_Result_base::_Deleter>,std::__future_base::_Task_state<marian::ThreadPool::enqueue<marian::data::BatchGenerator<marian::data::CorpusBase>::fetchBatchesAsync()::{lambda()#1}>(std::result_of&&,(marian::data::BatchGenerator<marian::data::CorpusBase>::fetchBatchesAsync()::{lambda()#1}&&)...)::{lambda()#1},std::allocator<int>,std::deque<std::shared_ptr<marian::data::CorpusBatch>,std::allocator<std::shared_ptr<marian::data::CorpusBatch>>> ()>::_M_run()::{lambda()#1},std::deque<std::shared_ptr<marian::data::CorpusBatch>,std::allocator<std::shared_ptr<marian::data::CorpusBatch>>>>>::  _M_invoke  (std::_Any_data const&) + 0x40
[0x560ec96ef139]    std::__future_base::_State_baseV2::  _M_do_set  (std::function<std::unique_ptr<std::__future_base::_Result_base,std::__future_base::_Result_base::_Deleter> ()>*,  bool*) + 0x29
[0x7f433d263827]                                                       + 0xf827
[0x560ec96f02aa]    std::__future_base::_Task_state<marian::ThreadPool::enqueue<marian::data::BatchGenerator<marian::data::CorpusBase>::fetchBatchesAsync()::{lambda()#1}>(std::result_of&&,(marian::data::BatchGenerator<marian::data::CorpusBase>::fetchBatchesAsync()::{lambda()#1}&&)...)::{lambda()#1},std::allocator<int>,std::deque<std::shared_ptr<marian::data::CorpusBatch>,std::allocator<std::shared_ptr>> ()>::  _M_run  () + 0x14a
[0x560ec96f11cb]    std::thread::_State_impl<std::thread::_Invoker<std::tuple<marian::ThreadPool::reserve(unsigned long)::{lambda()#1}>>>::  _M_run  () + 0x1ab
[0x7f433cf886df]                                                       + 0xbd6df
[0x7f433d25b6db]                                                       + 0x76db
[0x7f433c645a3f]    clone                                              + 0x3f
```

**Comment by tomsbergmanis** [comment]
Hi @pedrodiascoelho do you have any news with regards to if this is a fault in design or maybe just a bug that can be fixed?

**Comment by pedrodiascoelho** [comment]
Hi @tomsbergmanis 
I have not tested it my self, but I asked internally and i was told that experiments were already done with alignments and factors without any problem. The main difference is that they did not use the marian code after the changes that I've introduced, but used the regular marian repo, so I assume that the problem could be cause due to my implementation. Could you give more details regarding the options that you used regarding factors? If you used concatenation, I suggested to try without it and use the summing to run the default marian implementation (as I was told that that one worked with the alignments) while I try to figure out what introduced the problem.

**Comment by tomsbergmanis** [comment]
I did use concatenation, because summing in other experiments (without guided alignments) gave us worse performance. Is that something you have experienced as well? 

The other parameters are as follows: 
`$mariandir/marian --devices ${devices[@]} \
	--type transformer \
	--model $EXP_MODELS_SAVETO \
	--train-sets $EXP_TRAIN_SRC $EXP_TRAIN_TRG \
	--vocabs $EXP_DICT_SRC $EXP_DICT_TRG \
	--max-length 128 \
        --max-length-factor 1.5 \
	--mini-batch-fit \
	--workspace 9000 \
	--maxi-batch 200 \
	--early-stopping 40 \
	--valid-freq 1000 \
	--save-freq 2000 \
	--disp-freq 100 \
	--keep-best \
	--valid-metrics cross-entropy translation \
	--valid-sets $EXP_VALID_SRC $EXP_VALID_TRG \
	--valid-script-path $EXP_VALID_SCRIPT \
	--log $EXP_MODEL_DIR/train.log \
	--valid-log $EXP_MODEL_DIR/valid.log \
	--seed 347155 \
	--exponential-smoothing \
	--normalize 0.6 \
	--beam-size 12 \
	--quiet-translation \
	--valid-translation-output $EXP_MODEL_DIR/valid.output.txt \
	--valid-mini-batch 16 \
	--enc-depth 6 \
	--dec-depth 6 \
	--transformer-heads 8 \
	--transformer-preprocess d \
	--transformer-postprocess-emb d \
	--transformer-postprocess dan \
	--guided-alignment $EXP_ALIGNMENT \
	--transformer-guided-alignment-layer 1 \
	-guided-alignment-cost mse \
	--guided-alignment-weight 0.1 \
	--optimizer-delay $delay \
	--learn-rate 0.0005 \
	--lr-warmup 16000 \
	--lr-decay-inv-sqrt 16000 \
	--lr-report \
	--clip-norm 5 \
	--tied-embeddings-all \
	--sync-sgd \
	--transformer-dropout 0.1 \
	--transformer-dropout-attention 0.1 \
	--transformer-dropout-ffn 0.1 \
	--optimizer adam \
	--optimizer-params 0.9 0.98 1e-09 \
	--factors-combine concat \
	--factors-dim-emb 8
`

**Comment by pedrodiascoelho** [comment]
hi @tomsbergmanis, I tried training a simple model with your options and I was not able to duplicate your error. Everything worked fine for me when using guided alignments and concatenation chosen to combine lemma and factor embeddings as you did. Since you are using concat and --tied-embeddings-all, i assume you only used factors on the source side correct?

**Comment by tomsbergmanis** [comment]
That is right. I will try to look for an error on my end then. If anything I will keep you updated.

**Comment by tomsbergmanis** [comment]
It appears that the problem was on my side.
